### DISCUSSION TITLE
Error when trying to run a serverless function

### DISCUSSION THREAD

**Discussion by arthurvinci** [vercel/community#4895] [Help]
### Summary

Hi!
I am trying to make a very simple serverless function for my frontend code but I am struggling to make it work.
When deploying the code on Vercel and when I try to make the query, I get the error: 
`Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'express' imported from /var/task/api/index.js` 
and when I deploy it locally with `vercel dev` I get the error:
```
file:///usr/local/lib/node_modules/vercel/node_modules/@vercel/node/dist/dev-server.mjs:1016
    return listener(req, res);
           ^

TypeError: listener is not a function
    at Server.<anonymous> (file:///usr/local/lib/node_modules/vercel/node_modules/@vercel/node/dist/dev-server.mjs:1016:12)
```

Can someone help me fix that please?

### Steps to Reproduce

My `vercel.json` file is:
```json
{
  "rewrites": [{ "source": "/.well-known/(.*)", "destination": "/api" }]
}
```

And my `api` folder is an npm package with the following code in the `index.js`:
```js
import express from 'express';
import dapp_def from './radix.json' assert {type: 'json'} ;

const app = express()


app.get('/.well-known/radix.json', (req, res)=>{
    res.json(dapp_def)
})

app.listen(8081)
```

**Comment by eviltomato** [comment]
Same issue.

**Comment by eik-1** [comment]
Same issue
Were you able to resolve it ?

**Reply by kennybc** [reply]
Make sure to export your app object:
`export default app`

**Reply by Noumbissi16** [reply]
Yeah, it worked for me

I had the same error and just exporting app worked

**Comment by ZoomVirtuelst8** [comment]
same mistake 
Cannot find module '../../controller/controllerRegistros/cNuevoUserName.js'
Require stack:
- /var/task/src/routes/routesRegistros/rUserName.js
- /var/task/src/routes/index.js
- /var/task/src/server.js
- /var/task/index.js
- /var/task/___now_launcher.js
- /var/runtime/index.mjs
Did you forget to add it to "dependencies" in `package.json`?
INIT_REPORT Init Duration: 1013.05 ms	Phase: invoke	Status: error	Error Type: Runtime.ExitError
Error: Runtime exited with error: exit status 1

I have everything in order locally it works perfectly for me but in Vercel it doesn't work for me I've been working on it for a long time
![Screenshot 2024-03-31 062519](https://github.com/vercel/community/assets/156741405/7100f95e-bb6d-4c17-9c21-61c7d92d564f)
![Screenshot 2024-03-31 062117](https://github.com/vercel/community/assets/156741405/38d83148-e16e-4ed1-8a45-581f33a672a7)
![Screenshot 2024-03-31 061845](https://github.com/vercel/community/assets/156741405/cae701c0-80db-4080-9eaf-df0a43da0a16)
![Screenshot 2024-03-31 055817](https://github.com/vercel/community/assets/156741405/9ec67ebf-a5ee-47b5-8668-025da441fd5a)
![Screenshot 2024-03-31 054344](https://github.com/vercel/community/assets/156741405/f66aa533-5c1b-4504-9736-9f1d6e5151fb)
![Screenshot 2024-03-31 054146](https://github.com/vercel/community/assets/156741405/0275b049-43a2-47b6-ab46-5f320e5caf66)
![Screenshot 2024-03-31 054052](https://github.com/vercel/community/assets/156741405/00ac8a4b-ed11-4a2e-9a12-cfb25370cda9)
![Screenshot 2024-03-31 053748](https://github.com/vercel/community/assets/156741405/ee985fea-b4f8-41a6-8d86-d4f0ab652549)
![Screenshot 2024-03-31 053729](https://github.com/vercel/community/assets/156741405/2b0e1878-f984-4d55-af6e-4039219f1c0b)
![Screenshot 2024-03-31 053717](https://github.com/vercel/community/assets/156741405/3acb5e64-3a76-4c99-85d1-d494c4361635)
![Screenshot 2024-03-31 053047](https://github.com/vercel/community/assets/156741405/da98ff6e-3d46-4d55-944d-b2bc8794f6f6)

**Comment by ZoomVirtuelst8** [comment]
spanish
a mi me funcion usar este video https://www.youtube.com/watch?v=sO77i2VJyYE
y tambien usar   module.exports = server;
en el archivo index.js no se por que me pide exportar el server en index.js pero asi lo hice y ya me funciona 

english
Using this video worked for me https://www.youtube.com/watch?v=sO77i2VJyYE
and also use module.exports = server;
in the index.js file I don't know why it asks me to export the server in index.js but that's how I did it and it already works for me

index.js 
```
const server = require("./src/server.js");
const { conn } = require("./src/db.js");
const PORT = 3001;

conn
  .sync({ force: false })
  .then(async () => {
    server.listen(PORT || 3001, () => {
      console.log(`Server listening on port ${PORT}`);
    });
  })
  .catch((error) => console.error(error));
  module.exports = server;
```



server.js
```
const express = require("express");
require("dotenv").config();
const router = require("./routes/routes.js");
const morgan = require("morgan");
const cors = require("cors");
const helmet = require("helmet");
const server = express();
const corsConfig = {
  origin: "*",//"process.env.FRONT_URL",
  credential: true, // Habilita el envío de cookies y encabezados de autenticación
  methods: ["GET", "POST", "PUT", "DELETE"],
};

server.use(cors(corsConfig));
server.options("", cors(corsConfig));
server.use(express.json());
server.use(morgan("dev"));
server.use(helmet());

server.use(router);

module.exports = server;

```



vercel.json
```
{
  "version": 2,
  "builds": [
    {
      "src": "index.js",
      "use": "@now/node"
    }
  ],

  "routes": [
    {
      "src": "/(.*)",
      "dest": "index.js",
      "methods": ["GET", "POST", "PUT", "DELETE", "OPTIONS"]
    }
  ]
}

```

no se hasta donde les ayude pero eso fue lo que me ayudo 
I don't know how much it helped them but that's what helped me.

**Reply by eik-1** [reply]
I will try this and let you know

**Comment by akashdeepsinghal** [comment]
For Express in JS, in the main file, add the following line on the bottom:
`module.exports = app;`

For the given error, this should suffice.

However, for the Express in Typescipt project, as @kennybc suggested:
Add `export default app;` on bottom
And then build the app

**Reply by eik-1** [reply]
I'll try this and let you know

**Reply by eik-1** [reply]
This isn't working
Now the server just renders the index.js code

**Comment by Talalmahmud** [comment]
import express from "express";

const app = express();
const port = 5000
app.get("/", (req, res) => {
  res.status(200).json({ message: "Hello" });
});

app.listen((port), () => {
  console.log(`Server is running:http://localhost:${port}`);
});
export default app;
I used it and resolved my issue

**Comment by mozgbrasil** [comment]
I have the same problem using a project in Nest as per

https://github.com/orgs/vercel/discussions/7321
